### ISSUE TITLE
Request: Enable some kind of SSO functionality

### ISSUE THREAD

**Issue by alvitali** [Wingysam/Christmas-Community#131]
I'd love to tie Christmas-Community into my existing stack of selfhosted apps for family members and friends. Adding some sort of single sign-on with the option to create new users upon validation with an external auth service / database would go a long way to integrate this.

I'm aware this is probably way outside the scope of the current state of the project, but I'd love to see one of the following options:
- Support for LDAP to authenticate against an existing LDAP server
- Implementing OIDC with Christmas-Community as an OpenID Connect relying party

I'd love to help out with such a thing, but sadly have no experience in programming whatsoever. Maybe some day! In the meantime I want to thank you for building this app and making digital wishlists possible for our family.

**Comment by Wingysam** [comment]
How does #26 impact this issue?

**Comment by cj13579** [comment]
Given that the `local` package from Passport.js is already being used, it may not be too hard to configure some additional ones: [Passport.js strategies](https://www.passportjs.org/packages/).

**Comment by cj13579** [comment]
So I've had a go at this and got annoyingly close to getting it work with doing SSO to Google using the [passport-google-oidc](https://www.passportjs.org/packages/passport-google-oidc/) strategy in Passport.js. It authenticates fine, it even creates a new account if one doesn't already exist, but for some reason it won't login. I am sure I am doing something dumb. @Wingysam, perhaps you could take a peak at my branch: [here](https://github.com/Wingysam/Christmas-Community/compare/master...cj13579:Christmas-Community:google-oicd)?

You do need to register a client app in a Google project, but the instructions on the Passport.js page are pretty good in detailing what you need.

**Comment by Pr0mises** [comment]
> Given that the `local` package from Passport.js is already being used, it may not be too hard to configure some additional ones: [Passport.js strategies](https://www.passportjs.org/packages/).

Awesome work! Having OpenID and or OAuth2 implemented is highly appreciated

**Comment by maartenweyns** [comment]
I would also like the ability to put this behind a general SSO or OAuth2 provider, like Authentik for example. This makes managing users in a self hosted environment so much easier, both for me (the admin) and for my family members (the users).

**Comment by Eschguy** [comment]
I'll echo support for self-hosted SSO solutions like Authentik.

**Comment by cj13579** [comment]
I've spun up an instance of Authentik and I can take a look at doing an implementation for this at the weekend.

Do people feel that it would be OK to use the [Proxy Provider](https://docs.goauthentik.io/docs/add-secure-apps/providers/proxy/) and have Christmas-Community authenticate on the `X-Authentik-username` header and match that to the username in Christmas-Community?

**Comment by Eschguy** [comment]
Without knowing all the backend technical investment, I'd rather straight OIDC  support with something to allow user management to go through Authentik (and account creation on login if the account doesn't exist) is ideal.

**Comment by cj13579** [comment]
There is a bunch of clean-up and stuff missing (like docs and optional user creation) but I did get SSO working with Authentik using nginx proxy manager to reverse proxy to CC. I have a branch on my fork if people want to check it out.

@Wingysam - how do you feel about adding in (optional) support for automatic account creation when using an external authentication provider?

**Comment by Pr0mises** [comment]
> Without knowing all the backend technical investment, I'd rather straight OIDC  support with something to allow user management to go through Authentik (and account creation on login if the account doesn't exist) is ideal.

I can second that. Even though Auth Proxy is enough for Authentik users, it won't work for keycloak, Authelia etc as of my knowledge. OIDC is supported by the most SSO's

**Comment by cj13579** [comment]
I have created a PR for generic OIDC. It doesn't include "create a wishlist for this user if they don't exist" functionality though. That may best be handled under another issue/PR if that's something that people want/care about.

**Comment by Eschguy** [comment]
What happens if somebody logs in without a pre-made account? Are they just viewers or do they not get authenticated at all?

**Comment by cj13579** [comment]
> What happens if somebody logs in without a pre-made account? Are they just viewers or do they not get authenticated at all?

They don't get authenticated at all. You get an "[Unknown user](https://github.com/cj13579/Christmas-Community/blob/131-header-sso/src/languages/en-us.ts#L78)" error. 

Today with the Google integration (and going forward if/when #156 is merged) an admin creates a user account as normal and shares with the user the password creation/reset link. They can then login using their assigned username and password that they created. 

Once they are in the application, they can go to their profile and there is an option to link an SSO account. Clicking this button takes you off to the OIDC provider (Google, Authentik, whatever), you authenticate there and when it comes back to CC, we store your Authentication provider ID on your profile. 

If you then logout, when you log back in you can select the "Login with SSO" option from the login page. Clicking that will take you off to the OIDC provider and authenticate you. When you return back to CC, we get the ID from the provider and look up the account that's associated with that sign that account in.

**Comment by Eschguy** [comment]
Alright, good to know. Appreciate your work on this.

I think we should look into an auto-provisioning option once #156 is merged. For the sake of my family members, I'll definitely be doing all the login/link stuff before giving them info on the application.

**Comment by cj13579** [comment]
@brumik we don't have an issue for that yet, but I think that makes sense as a requirement for that issue!

**Comment by alvitali** [comment]
> I have created a PR for generic OIDC. It doesn't include "create a wishlist for this user if they don't exist" functionality though. That may best be handled under another issue/PR if that's something that people want/care about.

@cj13579 Thanks a lot for your work on the PR for OIDC support! It's great to have. For my personal use case however, automatic account creation for not yet existing users is a prerequisite for me to make full use of SSO. 

Reason being is that most of my family members / friends using Christmas Community already have an SSO account for my selfhosted apps, creating manual accounts for everybody and then linking them is quite the hurdle to adoption. What do others think about that?

I'd love to be able to implement this myself, however I lack most of the coding skills. Could look into doing this with ChatGPT though

**Comment by brumik** [comment]
I think we could make that update as a followup (see my group perms comment above). I am not entirely sure about the pace of this project but there are a lot of open prs that seem to be a bit slower side. @cj13579 Maybe you could say what is the eta on merging your PR as you seem to be a maintainer for a while?

**Comment by cj13579** [comment]
@Wingysam is the only one who has permission to merge and he seems to be very busy with college. 

To increase the velocity of changes additional maintainers could be added but ultimately that’s @Wingysam’s call.

**Comment by onlynow2** [comment]
> to their profile and there is an option to link an SSO account. Clicking this button takes you off to the OIDC provider (Google, Authentik, whatever), you a

Thank you, it helped me setting up SSO with Authentik as well. 
One thing I noticed was an issue with the "Redirect URIs/Origins" in Authentik.
Leaving this field empty helps, as Authentik is autofilling it with the correct values.

But when linking a SSO account the URI is different than with using SSO afterwards.
Therefore the following sequence was neccessary:
1. Not setting the Redirect URI
2. Linking an SSO account in CC (-> automatically inserted URI: https://cc.company/auth/oidc/link/redirect)
3. Deleting the Redirect URI again
4. Using the SSO in CC (-> automatically inserted URI: https://cc.company/auth/oidc/redirect)

That workaround is possible for my admin account, but not for every family member account.
Yet, I'm not quite sure where this issue origins from.

**Comment by cj13579** [comment]
@onlynow2 interesting, I'll see if I can replicate this later today!

**Comment by cj13579** [comment]
@onlynow2 - I _think_ I know what's going on here. You need to have both redirect URIs here. You add more than one by having them separated on multiple lines:

![Image](https://github.com/user-attachments/assets/ff99021d-9f38-4742-b41b-28c7ad69f6ce)

Below is some information on my Application setup in Authentik, for anyone else needing this in the future. In case its important for the Wizard flow stuff, I am running version 2024.10.2 of Authentik.

1. Create a new application:
    ![Image](https://github.com/user-attachments/assets/260d7282-a4bd-4379-9a02-96e16bd00439)
2. Select OAuth2/OIDC:
    ![Image](https://github.com/user-attachments/assets/83621d52-33b8-4ebe-af5e-22a83ab8b827)
3. Create a new provider using the "default-provider-authorization-implicit-consent". Add the two redirect URIs on separate lines. Copy your client ID and secret:
    ![Image](https://github.com/user-attachments/assets/dd076f5f-1a25-4d54-b181-4904902c9f0f)
4. You application creation is done:
    ![Image](https://github.com/user-attachments/assets/74e577b0-3166-475e-a3fb-2623f3756e66)
5. Setup your environment variables. For me I'm using a .env file but this might be on your docker-compose, k8s deployment file, whatever:
    ```
    OIDC_AUTHORIZATION_URL=https://authentik.cc/application/o/authorize/
    OIDC_TOKEN_URL=https://authentik.cc/application/o/token/
    OIDC_ISSUER=https://authentik.cc/application/o/onlynow2test/
    OIDC_CLIENT_ID=zk2feM1nM8Xa515KNkvH0LrVUPA5Df2lLDB4YDJZ
    OIDC_CLIENT_SECRET=REDACTED
    OIDC_PROVIDER_NAME=Authentik (onlynow2test)
    ```
6. Now, when I start the application I see this:
    ![Image](https://github.com/user-attachments/assets/7db2d317-7c84-4088-ac5e-57d33f0d6314)
7. Clicking the "Login with Authentik (onlynow2test)" I get redirected off to Authentik:
    ![Image](https://github.com/user-attachments/assets/cdd79fe6-404d-456f-9a70-122aa04dba69)
8. If I link an Account, I also see this now too:
    ![Image](https://github.com/user-attachments/assets/59f144f5-1ba3-44dd-a5de-2277cde45b15)

Hope that helps!

**Comment by onlynow2** [comment]
> [@onlynow2](https://github.com/onlynow2) - I _think_ I know what's going on here. You need to have both redirect URIs here. You add more than one by having them separated on multiple lines:
> 
> ![Image](https://github.com/user-attachments/assets/ff99021d-9f38-4742-b41b-28c7ad69f6ce)
> 
> Below is some information on my Application setup in Authentik, for anyone else needing this in the future. In case its important for the Wizard flow stuff, I am running version 2024.10.2 of Authentik.
> 
>     1. Create a new application:
>        ![Image](https://github.com/user-attachments/assets/260d7282-a4bd-4379-9a02-96e16bd00439)
> 
>     2. Select OAuth2/OIDC:
>        ![Image](https://github.com/user-attachments/assets/83621d52-33b8-4ebe-af5e-22a83ab8b827)
> 
>     3. Create a new provider using the "default-provider-authorization-implicit-consent". Add the two redirect URIs on separate lines. Copy your client ID and secret:
>        ![Image](https://github.com/user-attachments/assets/dd076f5f-1a25-4d54-b181-4904902c9f0f)
> 
>     4. You application creation is done:
>        ![Image](https://github.com/user-attachments/assets/74e577b0-3166-475e-a3fb-2623f3756e66)
> 
>     5. Setup your environment variables. For me I'm using a .env file but this might be on your docker-compose, k8s deployment file, whatever:
>        ```
>        OIDC_AUTHORIZATION_URL=https://authentik.cc/application/o/authorize/
>        OIDC_TOKEN_URL=https://authentik.cc/application/o/token/
>        OIDC_ISSUER=https://authentik.cc/application/o/onlynow2test/
>        OIDC_CLIENT_ID=zk2feM1nM8Xa515KNkvH0LrVUPA5Df2lLDB4YDJZ
>        OIDC_CLIENT_SECRET=REDACTED
>        OIDC_PROVIDER_NAME=Authentik (onlynow2test)
>        ```
> 
>     6. Now, when I start the application I see this:
>        ![Image](https://github.com/user-attachments/assets/7db2d317-7c84-4088-ac5e-57d33f0d6314)
> 
>     7. Clicking the "Login with Authentik (onlynow2test)" I get redirected off to Authentik:
>        ![Image](https://github.com/user-attachments/assets/cdd79fe6-404d-456f-9a70-122aa04dba69)
> 
>     8. If I link an Account, I also see this now too:
>        ![Image](https://github.com/user-attachments/assets/59f144f5-1ba3-44dd-a5de-2277cde45b15)
> 
> 
> Hope that helps!

Thanks for checking. I guess the pictured guide is pretty helpful for new users. 
Have you thought about referencing it in the documentation? :)

You're right, I did not think about putting both URIs there 😅 
I haven't had a service yet, which needed multiple URIs.

**Comment by cj13579** [comment]
> What is the correct format?

It depends. If you access your application at _https://wishlist.cc_ then you'll need to setup the following: 
- https://wishlist.cc/auth/oidc/redirect
- https://wishlist.cc/auth/oidc/link/redirect

**Comment by Rustymage** [comment]
> * /auth/oidc/redirect

Odd - with pocketid, neither work. I've tried both and individually. Only * works.

Thanks.

**Comment by tracygits** [comment]
Hi @cj13579 !

Really appreciate your visual guide on setting up OIDC. I followed all of your steps above but have been banging my head trying to figure out an error that I'm getting: "ID token not issued by expected OpenID provider." 

I can see in Authentik that I've been successfully authorized through, but it passes me that error after step 7.

- both redirect URIs are in there (passed that stage!).
- env. variables are similar with the trailing slash.
- binded existing user to the application in Authentik
- running Authentik version 2025.4.1

Any insights?

![Image](https://github.com/user-attachments/assets/0dafa82b-e90a-4970-9492-c9f5c8b56cee)

**Comment by cj13579** [comment]
@tracygits its not something I have seen before but you're likely going to want to check the value of the `OIDC_ISSUER` environment variable. It will need to be something like: `https://authentik.cc/application/o/{APPLICATION_NAME}/` replacing `{APPLICATION_NAME}` for the value of the _slug_ when you created the application.

**Comment by Pr0mises** [comment]
@cj13579 sure!
```
## Core Settings
# Where to store databases, can be a CouchDB compatible server or directory.
DB_PREFIX=dbs/
# Location of DB log file (if needed for debugging).
DB_LOG_FILE=/dev/null
# Where to send someone if they need to log in
DEFAULT_FAILURE_REDIRECT=/login
# Port to listen on
PORT=80
# Expose the internal PouchDB with CouchDB API and Fauxton browser. Mostly used for debugging. Leave empty to disable.
DB_EXPOSE_PORT=
# Proxy to send item data requests to. Leave empty to disable.
PROXY_SERVER=
# Secret string to store session cookies with. Automatically generated if not provided.
SECRET=
# How long a user is logged in (milliseconds). Defaults to one week.
SESSION_MAX_AGE=604800000
# The name of the site in the <title> and navigation bar
SITE_TITLE=Wishlist
# Used when shared to home screen
SHORT_TITLE=Wishlist
# The root URL for forms, CSS, and a small amount of JS. Useful when proxying or using SSO.
# If not using SSO, this can be a relative path.
ROOT_URL=/
# Where to trust the X-Forwarded-For header from. Defaults to "loopback". Useful for proxying to docker.
TRUST_PROXY=<redacted>
# Any theme from https://jenil.github.io/bulmaswatch
BULMASWATCH=default
# Set to false to disable update notices
UPDATE_CHECK=true
# Set to false to disable the profile pictures feature
PFP=true
# Language of the interface, options listed in `languages` directory
LANGUAGE=de-de
# Password to enter guest mode,
# e.g. https://wishes.example.com?pw=ReplaceWithYourGuestPassword
# GUEST_PASSWORD=ReplaceWithYourGuestPassword

## Wishlist Settings
# Set to true to not allow users to have their own lists. You may want this for a birthday or wedding.
SINGLE_LIST=true
# Set to false to allow viewing wishlists without logging in
LISTS_PUBLIC=true
# Defaults to true. Set to false for legacy cards view.
TABLE=true
# Allow Markdown in item notes. Does not work with TABLE=false. Defaults to false.
MARKDOWN=true

## Custom HTML Snippets
# These are inserted into specific locations in the relevant page
# HTML is not escaped. Don't put untrusted data here.
# CUSTOM_HTML_LOGIN=<p style="margin-top: 1em;">Some custom text for the Login page</p>
# CUSTOM_HTML_WISHLISTS=

# Custom CSS stylesheet
# If you wish to include a custom stylesheet you can add the filename in the variable here.
# Remember to add the stylesheet to the filesystem at `static/css/custom.css`. In docker, mount `/usr/src/app/src/static/css/custom.css`.
# CUSTOM_CSS=custom.css

## Google Client Details
# You can configure single sign-on to your Christmas Community instance using Google accounts. Read this guide for details of what to configure on the Google side: https://developers.google.com/identity/openid-connect/openid-connect
# Once you've created a client ID and secret in your Google project use the below environment variables to enable SSO

# GOOGLE_CLIENT_ID=
# GOOGLE_CLIENT_SECRET=

## OIDC Provider Details
# You can configure single sign-on to your Christmas Community instance using any OIDC provider.
# Once you've created a client ID and secret in your authentication provider use the below environment variables to enable single sign on.
# The values specified below are the defaults
OIDC_CLIENT_ID=<redacted>
OIDC_CLIENT_SECRET=<redacted>
OIDC_AUTHORIZATION_URL=https://auth.domain.comn/application/o/authorize/
OIDC_TOKEN_URL=https://auth.domain.com/application/o/token/
OIDC_ISSUER=https://auth.domain.com/application/o/wishlist/
OIDC_PROVIDER_NAME=Authentik
OIDC_AUTO_CREATE_USER=true


# Profile picture upload max size in MB
UPLOAD_PFP_MAX_SIZE=5

```
